File: _site/solutions/question_82_remove_duplicats_sorted_list_2.rb

# Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.
#
# For example,
# Given 1->2->3->3->4->4->5, return 1->2->5.
# Given 1->1->1->2->3, return 2->3.

# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
  def initialize(val)
    @val = val
    @next = nil
  end
end

# @param {ListNode} head
# @return {ListNode}
def delete_duplicates(head)
  r = Hash.new(0)
  while head
    r[head.val] += 1
    head = head.next
  end

  head = dummy = ListNode.new(-1)

  r.each_pair do |k, v|
    next if v > 1
    head.next = ListNode.new(k)
    head = head.next
  end
  dummy.next
end

list = ListNode.new(1)
list.next = ListNode.new(1)
list.next.next = ListNode.new(1)
list.next.next.next = ListNode.new(2)
list.next.next.next.next = ListNode.new(2)
list.next.next.next.next.next = ListNode.new(3)

p delete_duplicates(list)

File: _site/solutions/question_82_remove_duplicats_sorted_list_2.py
from collections import OrderedDict
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        r = OrderedDict()
        while head:
            v = head.val
            r[v] = r[v] + 1 if v in r else 1
            head = head.next

        dummy = head = ListNode(-1)
        for k, v in r.items():
            if v == 1:
                head.next = ListNode(k)
                head = head.next
        return dummy.next
